### PR TITLE
feat(chart): Add optional env and envFrom parameters for all deployments and cronjobs

### PR DESCRIPTION
**Pull request by MPritsch**
Adds the possibility to add env values or envFrom values for all deployments and cronjobs. We use kargo on a cluster behind a webproxy and needed a way to configure http_proxy and no_proxy variables in the kargo-controller. Currently we work around this by patching the deployment with kustomize. With this PR it can be done via helm-values.

While I technically only need these values on the kargo-controller, I wanted to add the env and envFrom values to all deployments to keep it consistent.

I tried to stay close to how argocd lets you configure these values in their helm charts. https://github.com/argoproj/argo-helm/tree/main/charts/argo-cd

Used this values file locally to check that the functionality works:
```yaml
global:
  env:
    - name: GLOBAL_ENV
      value: "========="

api:
  env:
    - name: EXTRA_ENV_API
      value: "-----------------"
  envFrom:
    - configMapRef:
        name: EXTRA_CM_API
  oidc:
    enabled: true
    dex:
      enabled: true
      env:
        - name: EXTRA_ENV_DEX
          value: "-----------------"
      envFrom:
        - configMapRef:
            name: EXTRA_CM_DEX
  adminAccount:
    passwordHash: $2a$12$qG6iGJgTpQbZo3dRi3q1a.gYg8Na8PF8f263VcdcQNN.squppsuSG
    tokenSigningKey: iwishtowashmyirishwristwatch

controller:
  env:
    - name: EXTRA_ENV_CONTROLLER
      value: "-----------------"
  envFrom:
    - configMapRef:
        name: EXTRA_CM_CONTROLLER

garbageCollector:
  env:
    - name: EXTRA_ENV_GARBAGE_COLLECTOR
      value: "-----------------"
  envFrom:
    - configMapRef:
        name: EXTRA_ENV_GARBAGE_COLLECTOR

managementController:
  env:
    - name: EXTRA_ENV_MANAGEMENT_CONTROLLER
      value: "-----------------"
  envFrom:
    - configMapRef:
        name: EXTRA_ENV_MANAGEMENT_CONTROLLER

webhooksServer:
  env:
    - name: EXTRA_ENV_WEBHOOK
      value: "-----------------"
  envFrom:
    - configMapRef:
        name: EXTRA_CM_WEBHOOK
```

With this command:
```yaml
helm template kargo . --namespace kargo --create-namespace  -f test.values.yaml > test.yaml
```

Best regards. Thanks for your work.

### CODE REVIEW

File: charts/kargo/templates/dex-server/deployment.yaml [hunk 1]
```diff
@@ -44,6 +44,10 @@ spec:
         {{- with .Values.api.oidc.dex.env }}
```
Comment (krancour) [inline charts/kargo/templates/dex-server/deployment.yaml:44]:
```suggestion
        {{- with concat (.Values.global.env .Values.api.oidc.dex.env) }}
```

File: charts/kargo/values.yaml [hunk 2]
```diff
@@ -3,6 +3,13 @@
 ## This is a YAML-formatted file.
 ## Declare variables to be passed into your templates.
 
+## @section Global Parameters
+global:
+  ## @param global.env env to pass to all deployed Deployments and CronJobs
```
Comment (krancour) [inline charts/kargo/values.yaml:8]:
```suggestion
  ## @param global.env Supplemental environment variables for all Kargo workloads.
```

File: charts/kargo/values.yaml [hunk 3]
```diff
@@ -88,6 +95,16 @@ api:
   annotations: {}
   ## @param api.securityContext Security context for api pods.
   securityContext: {}
+  ## @param api.env env to pass to api pods.
```
Comment (krancour) [inline charts/kargo/values.yaml:98]:
```suggestion
  ## @param api.env Supplemental environment variables for API server pods.
```

File: charts/kargo/values.yaml [hunk 4]
```diff
@@ -88,6 +95,16 @@ api:
...
+  env: []
+    #  - name: GLOBAL_ENV
+    #    value: =========
+  ## @param api.envFrom envFrom to pass to api pods.
```
Comment (krancour) [inline charts/kargo/values.yaml:102]:
```suggestion
  ## @param api.envFrom Supplemental environment variables for API server pods from from ConfigMaps or Secrets.
```

File: charts/kargo/values.yaml [hunk 5]
```diff
@@ -195,6 +212,12 @@ api:
       #     secretKeyRef:
       #       name: github-dex
       #       key: dex.github.clientSecret
+      ## @param api.oidc.dex.envFrom envFrom to pass to Dex pods.
```
[the remark this answers is not in the document]
Comment (krancour) [reply charts/kargo/values.yaml:215]: I think we should also change `api.oidc.dex.env` a few lines up to:

```
@param api.oidc.dex.env Supplemental environment variables for Dex server pods. This is especially convenient for cases where api.oidc.dex.connectors needs to reference environment variables from a Secret that is managed "out of band" with a secret management solution such as Sealed Secrets.
```

File: charts/kargo/values.yaml [hunk 6]
```diff
@@ -336,6 +359,16 @@ controller:
   affinity: {}
   ## @param controller.annotations Annotations to add to the controller pods.
   annotations: {}
+  ## @param controller.env env to pass to controller pods.
```
Comment (krancour) [inline charts/kargo/values.yaml:362]:
```suggestion
  ## @param controller.env Supplemental environment variables for controller pods.
```

File: charts/kargo/values.yaml [hunk 7]
```diff
@@ -336,6 +359,16 @@ controller:
...
+  env: []
+    #  - name: GLOBAL_ENV
+    #    value: =========
+  ## @param controller.envFrom envFrom to pass to controller pods.
```
Comment (krancour) [inline charts/kargo/values.yaml:366]:
```suggestion
  ## @param controller.envFrom envFrom Supplemental environment variables for controller pods from from ConfigMaps or Secrets.
```

File: charts/kargo/values.yaml [hunk 8]
```diff
@@ -365,6 +398,16 @@ managementController:
   affinity: {}
   ## @param managementController.annotations Annotations to add to the management controller pods.
   annotations: {}
+  ## @param managementController.env env to pass to management controller pods.
```
Comment (krancour) [inline charts/kargo/values.yaml:401]:
```suggestion
  ## @param managementController.env Supplemental environment variables for management controller pods.
```

File: charts/kargo/values.yaml [hunk 9]
```diff
@@ -365,6 +398,16 @@ managementController:
...
+  env: []
+    #  - name: GLOBAL_ENV
+    #    value: =========
+  ## @param managementController.envFrom envFrom to pass to management controller pods.
```
Comment (krancour) [inline charts/kargo/values.yaml:405]:
```suggestion
  ## @param managementController.envFrom Supplemental environment variables for management controller pods from ConfigMaps or Secrets.
```

File: charts/kargo/values.yaml [hunk 10]
```diff
@@ -403,6 +446,16 @@ webhooksServer:
   annotations: {}
   ## @param webhooksServer.securityContext Security context for webhooks server pods.
   securityContext: {}
+  ## @param webhooksServer.env env to pass to webhooks server pods.
+  env: []
+    #  - name: GLOBAL_ENV
+    #    value: =========
+  ## @param webhooksServer.envFrom envFrom to pass to webhooks server pods.
```
Comment (krancour) [inline charts/kargo/values.yaml:453]:
```suggestion
  ## @param webhooksServer.envFrom Supplemental environment variables for webhook server pods from ConfigMaps or Secrets.
```

File: charts/kargo/values.yaml [hunk 11]
```diff
@@ -403,6 +446,16 @@ webhooksServer:
   annotations: {}
   ## @param webhooksServer.securityContext Security context for webhooks server pods.
   securityContext: {}
+  ## @param webhooksServer.env env to pass to webhooks server pods.
```
Comment (krancour) [inline charts/kargo/values.yaml:449]:
```suggestion
  ## @param webhooksServer.env Supplemental environment variables for webhook server pods.
```

File: charts/kargo/values.yaml [hunk 12]
```diff
@@ -440,3 +493,13 @@ garbageCollector:
   annotations: {}
   ## @param garbageCollector.securityContext Security context for garbage collector pods.
   securityContext: {}
+  ## @param garbageCollector.env env to pass to garbage collector pods.
```
Comment (krancour) [inline charts/kargo/values.yaml:496]:
```suggestion
  ## @param garbageCollector.env Supplemental environment variables for garbage collector pods.
```

File: charts/kargo/values.yaml [hunk 13]
```diff
@@ -440,3 +493,13 @@ garbageCollector:
...
+  env: []
+    #  - name: GLOBAL_ENV
+    #    value: =========
+  ## @param garbageCollector.envFrom envFrom to pass to garbage collector pods.
```
Comment (krancour) [inline charts/kargo/values.yaml:500]:
```suggestion
  ## @param garbageCollector.envFrom Supplemental environment variables for garbage collector pods from ConfigMaps or Secrets.
```

File: charts/kargo/values.yaml [hunk 14]
```diff
@@ -3,6 +3,13 @@
 ## This is a YAML-formatted file.
 ## Declare variables to be passed into your templates.
 
+## @section Global Parameters
+global:
+  ## @param global.env env to pass to all deployed Deployments and CronJobs
+  env: []
+    #  - name: GLOBAL_ENV
+    #    value: =========
```
Comment (krancour) [inline charts/kargo/values.yaml:11]:
```suggestion
  env: []
  #  - name: ENV_NAME
  #    value: value
```

File: charts/kargo/values.yaml [hunk 15]
```diff
@@ -88,6 +95,16 @@ api:
   annotations: {}
   ## @param api.securityContext Security context for api pods.
   securityContext: {}
+  ## @param api.env env to pass to api pods.
+  env: []
+    #  - name: GLOBAL_ENV
+    #    value: =========
```
Comment (krancour) [inline charts/kargo/values.yaml:101]:
```suggestion
  env: []
  #  - name: ENV_NAME
  #    value: value
```

File: charts/kargo/values.yaml [hunk 16]
```diff
@@ -336,6 +359,16 @@ controller:
   affinity: {}
   ## @param controller.annotations Annotations to add to the controller pods.
   annotations: {}
+  ## @param controller.env env to pass to controller pods.
+  env: []
```
Comment (krancour) [inline charts/kargo/values.yaml:363]:
```suggestion
  env: []
  #  - name: ENV_NAME
  #    value: value
```

File: charts/kargo/values.yaml [hunk 17]
```diff
@@ -336,6 +359,16 @@ controller:
...
+    #  - name: GLOBAL_ENV
+    #    value: =========
```
Comment (krancour) [inline charts/kargo/values.yaml:365]:
```suggestion
  env: []
  #  - name: ENV_NAME
  #    value: value
```

File: charts/kargo/values.yaml [hunk 18]
```diff
@@ -365,6 +398,16 @@ managementController:
   affinity: {}
   ## @param managementController.annotations Annotations to add to the management controller pods.
   annotations: {}
+  ## @param managementController.env env to pass to management controller pods.
+  env: []
+    #  - name: GLOBAL_ENV
+    #    value: =========
```
Comment (krancour) [inline charts/kargo/values.yaml:404]:
```suggestion
  env: []
  #  - name: ENV_NAME
  #    value: value
```

File: charts/kargo/values.yaml [hunk 19]
```diff
@@ -403,6 +446,16 @@ webhooksServer:
   annotations: {}
   ## @param webhooksServer.securityContext Security context for webhooks server pods.
   securityContext: {}
+  ## @param webhooksServer.env env to pass to webhooks server pods.
+  env: []
+    #  - name: GLOBAL_ENV
+    #    value: =========
```
Comment (krancour) [inline charts/kargo/values.yaml:452]:
```suggestion
  env: []
  #  - name: ENV_NAME
  #    value: value
```

File: charts/kargo/values.yaml [hunk 20]
```diff
@@ -440,3 +493,13 @@ garbageCollector:
   annotations: {}
   ## @param garbageCollector.securityContext Security context for garbage collector pods.
   securityContext: {}
+  ## @param garbageCollector.env env to pass to garbage collector pods.
+  env: []
+    #  - name: GLOBAL_ENV
+    #    value: =========
```
Comment (krancour) [inline charts/kargo/values.yaml:499]:
```suggestion
  env: []
  #  - name: ENV_NAME
  #    value: value
```

File: charts/kargo/values.yaml [hunk 21]
```diff
@@ -88,6 +95,16 @@ api:
...
+  ## @param api.env env to pass to api pods.
+  env: []
+    #  - name: GLOBAL_ENV
+    #    value: =========
+  ## @param api.envFrom envFrom to pass to api pods.
+  envFrom: []
+    #  - configMapRef:
+    #      name: extra-configmap
+    #  - secretRef:
+    #      name: extra-secret
```
Comment (krancour) [inline charts/kargo/values.yaml:107]:
```suggestion
  envFrom: []
  #  - configMapRef:
  #      name: config-map-name
  #  - secretRef:
  #      name: secret-name
```

File: charts/kargo/values.yaml [hunk 22]
```diff
@@ -195,6 +212,12 @@ api:
       #     secretKeyRef:
       #       name: github-dex
       #       key: dex.github.clientSecret
+      ## @param api.oidc.dex.envFrom envFrom to pass to Dex pods.
+      envFrom: []
+        #  - configMapRef:
+        #      name: extra-configmap
+        #  - secretRef:
+        #      name: extra-secret
```
Comment (krancour) [inline charts/kargo/values.yaml:220]:
```suggestion
  envFrom: []
  #  - configMapRef:
  #      name: config-map-name
  #  - secretRef:
  #      name: secret-name
```

File: charts/kargo/values.yaml [hunk 23]
```diff
@@ -336,6 +359,16 @@ controller:
...
+  ## @param controller.env env to pass to controller pods.
+  env: []
+    #  - name: GLOBAL_ENV
+    #    value: =========
+  ## @param controller.envFrom envFrom to pass to controller pods.
+  envFrom: []
+    #  - configMapRef:
+    #      name: extra-configmap
+    #  - secretRef:
+    #      name: extra-secret
```
Comment (krancour) [inline charts/kargo/values.yaml:371]:
```suggestion
  envFrom: []
  #  - configMapRef:
  #      name: config-map-name
  #  - secretRef:
  #      name: secret-name
```

File: charts/kargo/values.yaml [hunk 24]
```diff
@@ -365,6 +398,16 @@ managementController:
...
+  ## @param managementController.env env to pass to management controller pods.
+  env: []
+    #  - name: GLOBAL_ENV
+    #    value: =========
+  ## @param managementController.envFrom envFrom to pass to management controller pods.
+  envFrom: []
+    #  - configMapRef:
+    #      name: extra-configmap
+    #  - secretRef:
+    #      name: extra-secret
```
Comment (krancour) [inline charts/kargo/values.yaml:410]:
```suggestion
  envFrom: []
  envFrom: []
  #  - configMapRef:
  #      name: config-map-name
  #  - secretRef:
  #      name: secret-name
```

File: charts/kargo/values.yaml [hunk 25]
```diff
@@ -403,6 +446,16 @@ webhooksServer:
...
+  ## @param webhooksServer.env env to pass to webhooks server pods.
+  env: []
+    #  - name: GLOBAL_ENV
+    #    value: =========
+  ## @param webhooksServer.envFrom envFrom to pass to webhooks server pods.
+  envFrom: []
+    #  - configMapRef:
+    #      name: extra-configmap
+    #  - secretRef:
+    #      name: extra-secret
```
Comment (krancour) [inline charts/kargo/values.yaml:458]:
```suggestion
  envFrom: []
  #  - configMapRef:
  #      name: config-map-name
  #  - secretRef:
  #      name: secret-name
```

File: charts/kargo/values.yaml (hunk 2 above)
Comment (krancour) [reply charts/kargo/values.yaml:8]: I was also wondering why no `globa.envFrom`?

Comment (MPritsch) [reply charts/kargo/values.yaml:8]: Just because I stuck to the argoCD helm template where there was no global.envFrom. We can add it if you want to.

**Comment by krancour** [comment]
@MPritsch this is a sensible and welcome change. Thank you!

I made several suggestions, mostly about formatting of values.yaml, just to promote consistency with some other, existing settings. (Unfortunately, these will require re-running codegen. Sorry!)

I did have one question about why no global `envFrom`.

These minor things aside, this looks good and I think we should be able to get this into v0.6.0.

**Comment by MPritsch** [comment]
@krancour I applied your suggested changes. Let me know if you want to also add the `global.envFrom`. Like I explained, the only reason I didn't add it, because argoCD's helmChart also didn't use it.

**Comment by krancour** [comment]
> the only reason I didn't add it, because argoCD's helmChart also didn't use it.

Sorry. I overlooked that detail.

I'd still like to add it. Argo CD is in our DNA, but we're not shy about deviating from it if something makes sense to us.

**Comment by MPritsch** [comment]
> I'd still like to add it. Argo CD is in our DNA, but we're not shy about deviating from it if something makes sense to us.

Added global.envFrom

**Comment by krancour** [comment]
@MPritsch this is really a great change and fantastically executed. Thank you!

**Review by krancour — APPROVED** [review]

**Pull request merged**